### PR TITLE
Upgrade to PTL 2.1 and 2.2

### PR DESCRIPTION
**Pull request by athitten**
# What does this PR do ?

Upgrades NeMo to use PTL 2.1 and 2.2. NeMo now uses PTL 2.2.1 (the current latest release).

Major changes that are seen across a lot of files in the PR:

1. The function signature of `train/validation/test_step` is changed while using the `dataloader_iter` API. The `batch_idx` arg is removed from the function signature and `next(dataloader_iter)` now returns `batch, batch_idx, dataloader_idx`. 

With this change the support for multiple dataloaders which was removed with PTL 2.0 is added back. [More details here](https://lightning.ai/docs/pytorch/stable/upgrade/from_2_0.html#:~:text=predict_step(self%2C%20dataloader_iter%2C%20batch_idx)%20in%20LightningModule-,remove%20batch_idx%20from%20the%20signature%20and%20expect%20dataloader_iter%20to%20return%20a%20triplet%20(batch%2C%20batch_idx%2C%20dataloader_idx),-PR18390)

2. PTL 2.1 and above does not support having both precision and  precision plugin in the trainer. It can be observed in the changes that if precision plugin is created, then precision is made None.

3. Some of the other simple changes are explained with comments in the PR.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Jenkins CI
To run Jenkins, a NeMo User with write access must comment `jenkins` on the PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

File: nemo/collections/multimodal/models/multimodal_llm/neva/neva_model.py [hunk 1]
```diff
@@ -613,16 +613,16 @@ def forward(self, tokens, text_position_ids, attention_mask, labels, media=None)
...
         return output_tensor
 
-    def fwd_bwd_step(self, dataloader_iter, batch_idx, forward_only, first_val_step=None):
-        return MegatronGPTModel.fwd_bwd_step(self, dataloader_iter, batch_idx, forward_only, first_val_step)
+    def fwd_bwd_step(self, dataloader_iter, forward_only, first_val_step=None):
+        return MegatronGPTModel.fwd_bwd_step(self, dataloader_iter, forward_only, first_val_step)
 
-    def training_step(self, dataloader_iter, batch_idx):
+    def training_step(self, dataloader_iter):
         """
```
Comment (athitten) [inline nemo/collections/multimodal/models/multimodal_llm/neva/neva_model.py:620]: PTL 2.1 introduces change in the training/validation/test_step signature while using `dataloader_iter` API. Removes `batch_idx` from the signature and `dataloader_iter` returns a triplet` (batch, batch_idx, dataloader_idx)`: [More details here](https://lightning.ai/docs/pytorch/stable/upgrade/from_2_0.html#:~:text=training_step(self%2C%20dataloader_iter%2C%20batch_idx)%20in%20LightningModule-,remove%20batch_idx%20from%20the%20signature%20and%20expect%20dataloader_iter%20to%20return%20a%20triplet%20(batch%2C%20batch_idx%2C%20dataloader_idx),-PR18390)

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 2]
```diff
@@ -1125,17 +1138,13 @@ def id_func(output_tensor):
...
         """
             Our dataloaders produce a micro-batch and then we fetch
             a number of microbatches depending on the global batch size and model parallel size
             from the dataloader to produce a list of microbatches.
             The list of microbatches is then piped through the pipeline using megatron-core fwd/bwd functions.
         """
-        # Check if iterator is exhausted
-        dataloader_iter, done = self._val_iterator_done(dataloader_iter)
-        if done:
-            return
```
Comment (athitten) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:1138]: This check is not required anymore since PTL takes care of it.

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

**Comment by athitten** [comment]
jenkins

File: requirements/requirements_lightning.txt [hunk 3]
```diff
@@ -1,6 +1,6 @@
 hydra-core>1.3,<=1.3.2
 omegaconf<=2.3
-pytorch-lightning>=2.0,<=2.0.7
+pytorch-lightning==2.2.1
```
Comment (ericharper) [inline requirements/requirements_lightning.txt:3]: should be >= 2.2.1

**Comment by ericharper** [comment]
jenkins

File: requirements/requirements_lightning.txt (hunk 3 above)
Comment (athitten) [reply requirements/requirements_lightning.txt]: Fixed in the latest commit. @ericharper let me know if you have any other review comments. The CI passed fully and we should be good to merge after the code review.

**Comment by athitten** [comment]
@yaoyu-33 @Victor49152 can one of you all review and approve the PR from the multi modal side ? Thank you!

**Comment by athitten** [comment]
jenkins

**Review by Victor49152 — APPROVED** [review]
LGTM on multimodal side

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py [hunk 4]
```diff
@@ -18,6 +18,7 @@
 
 import torch
 from omegaconf import DictConfig, ListConfig
+from pytorch_lightning.loops.fetchers import _DataFetcherWrapper
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:21]: Just noting here we're using an `_` class. Do we need to ask PTL for an API for this?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py [hunk 5]
```diff
@@ -460,7 +456,7 @@ def inference_step(self, dataloader_iter, batch_idx, mode, dataloader_idx=0):
 
     def inference_epoch_end(self, outputs, mode, data_cfg):
         # Parent class will handle logging of the loss.
-        if not outputs:
+        if not outputs or not outputs[0]:
```
Comment (ericharper) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:459]: This seems risky as well. Can we ask PTL for an API for outputs? It seems we never know what we're getting with it.

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

Left a couple of comments as follow ups for PTL.

**Pull request merged**